lib/record_store/provider/oracle_cloud_dns.rb

```diff
@@ -0,0 +1,182 @@
+require 'oci'
+
+module RecordStore
+  class Provider::OracleCloudDNS < Provider
+    class Error < StandardError; end
+
+    class << self
+      def client
+        @client ||= begin
+          config = OCI::Config.new
+          config.user = secrets['user']
+          config.fingerprint = secrets['fingerprint']
+          config.key_content = secrets['key_content']
+          config.tenancy = secrets['tenancy']
+          config.region = secrets['region']
+          config.validate
+          OCI::Dns::DnsClient.new(config: config)
+        end
+      end
+
+      # Downloads all the records from the provider.
+      #
+      # Returns: an array of `Record` for each record in the provider's zone
+      def retrieve_current_records(zone:, stdout: $stdout)
+        client.get_zone_records(zone)
+          .flat_map { |response| response.data.items }
+          .flat_map { |api_record| build_from_api(api_record) }
+          .compact
+      rescue OCI::Errors::HttpRequestBasedError
+        stdout.puts "Cannot build record for zone: #{zone}"
+        raise
+      end
+
+      # Returns an array of the zones managed by provider as strings
+      def zones
+        client.list_zones(secrets['compartment_id']).data.map(&:name)
+      end
+
+      private
+
+      # Creates a new record to the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def add(record, zone)
+        patch_add_record = [
+          OCI::Dns::Models::RecordOperation.new(
+            domain: record.fqdn,
+            rtype: record.type,
+            ttl: record.ttl,
+            rdata: record.rdata_txt,
+            operation: 'ADD',
+          ),
+        ]
+
+        client.patch_zone_records(
+          zone,
+          OCI::Dns::Models::PatchZoneRecordsDetails.new(items: patch_add_record)
+        )
+      end
+
+      # Deletes an existing record from the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def remove(record, zone)
+        record_fqdn = Record.ensure_ends_without_dot(record.fqdn)
+        found_record = client.get_zone_records(
+          zone,
+          rtype: record.type,
+          zone_version: client.get_zone(zone).data.version,
+          domain: record_fqdn,
+        ).data.items.last
```

Comment:
It seems like we're still fetching an arbitrary (always the last) record to model our deleted record off of.  Consider two `TXT` records on a single FQDN:

```
- fqdn: foo.com.
  type: TXT
  record_hash: 0
  rrdata: text1
- fqdn: foo.com.
  type: TXT
  record_hash: 1
  rrdata: text2
```

...and let's say `remove()` is called on the first record (`text1`).  Wouldn't the `record_hash` we use on line 80 always be that of the last record retrieved (`text2`)?  Would the patch request work out?

Reply:
It worked out with Cassette test cases regardless of record's types

Reply:
Would it be possible to add parameters to fetch the record ?
Or maybe use `select` on the returned records instead of `last` to make sure the one we are removing is the one we were supposed to remove ?

In your tests, have you tried deleting a record, which is not the last one, while there is another record on the exact same domain of the exact same type ? It seems error-prone to me too, as @cttttt said

Reply:
@XaF  Chris and I just have discussed about it and it deletes first one even though I create two records under same `fqdn`. In case it might not work with other types, I tested 'A' and 'TXT'. Both deleted the first one. But I've been working on adding condition to make sure `rdata` is also same or not. Will merge & ship ip first and watch how it works with `record-store`.

Reply:
It worked as expected but it still seems odd so I updated the code to check `r.data == record.rdata_type`.